### ISSUE TITLE
Variable Explorer gets busy with large variables

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#572]
_From [gestat...@gmail.com](https://code.google.com/u/114884686971745659534/) on 2011-03-08T10:08:42Z_

What steps will reproduce the problem?
1. Open the variable explorer
   1. Load some couple hundred megabyte variables into the current interpreter

The variable explorer gets so busy updating it's data that the whole spyder gets really laggy. The python interpreter spyder is running in, is occupied literally two thirds of the time (see attached load graphic).

I found out when I detached the variable explorer so it had to update it's data all the time. 

I'm using
Python 2.6.5
Qt 4.6.2
PyQt 4.7.2
on x86_64 GNU/Linux

**Attachment:** [load.png](http://code.google.com/p/spyderlib/issues/detail?id=572)

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=572_

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-03-08T09:59:35Z_

The variable explorer is being refreshed periodically, hence this periodic CPU occupation. The timeout may be customized and this automatic refresh feature may even be disabled if necessary (see Preferences > Variable explorer).

Furthermore, the CPU load depends on the Variable explorer options: computing min/max for arrays takes time, representing these arrays can take time too, and so on. Try to play with the display options (which are both in Variable explorer's vertical toolbar and in Preferences dialog box) to see if there is an improvement.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-03-08T10:01:08Z_

Oh and could you please be more specific about the number of loaded arrays and about their sizes?

I'm working with large arrays quite often and I did not experience such behavior.

**Comment by spyder-issues-migrator** [comment]
_From [gestat...@gmail.com](https://code.google.com/u/114884686971745659534/) on 2011-03-09T00:33:24Z_

I'm talking about approximately 20 to 30 float64 numpy arrays with a couple of megabytes each. The total amount of memory would be approx 200 MB.

min/max calculation is disabled. Increasing the update interval of course alleviates the effect but doesn't really solve the problem. I decided to disable the periodic update at all for the time being.

It seems like the effect doesn't show all the time. A requirement for reproducing this phenomenon, as it seems, is to place the variable explorer in a dedicated window tile such that it is always visible. Whenever the variable explorer shared a tile with other elements, the problem didn't show up.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-03-09T09:36:04Z_

Note that the Variable explorer is updated only if visible (so if the dockwidget is tabified with another dockwidget and hidden, no refresh is the expected behavior).

Try to reproduce the issue like this:
(because I'm trying to reproduce your problem that way but I can't)

for i in range(20):
    globals()['x%d' % i] = np.linspace(-10, 10, 300000)

**Comment by spyder-issues-migrator** [comment]
_From [gestat...@gmail.com](https://code.google.com/u/114884686971745659534/) on 2011-03-10T01:02:26Z_

Same here. Cannot reproduce this using your code. The glitch seems to be caused by something I do in my simulation script. I'll keep trying to pinpoint the origin. 

I'll keep you posted.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-05-16T01:28:53Z_

**Labels:** Cat-VariableExplorer

**Comment by spyder-issues-migrator** [comment]
_From [kinnison...@gmail.com](https://code.google.com/u/104685941974947235171/) on 2013-04-29T12:05:25Z_

I've had plenty of large variables before without too much of a problem, but recently ran into severe slowdowns related to the variable explorer.  The slowdown wasn't periodic, but instead occurred only when I tried to access some data.  Even unrelated data from a simple list of integers.  Problem goes away when I delete the large objects.  But it also goes away if I close the variable explorer entirely.  No variable explorer settings helped (and I disabled updating well before this problem).

I'll see if I can get a reproduceable example.  I can reproduce the problem on my own, but need to figure out how to get the data in a sharable format.

**Comment by spyder-issues-migrator** [comment]
_From [kinnison...@gmail.com](https://code.google.com/u/104685941974947235171/) on 2013-04-29T13:20:30Z_

I think I figured it out.

class TakesLongToRepr(object):
    def __repr__(self):
        print id(self)
        return 'hi'
trouble=np.array([TakesLongToRepr() for _ in xrange(1)])
y=range(50)

Run that, then start accessing the values in y from the console.  y[3], then y[40], etc.  Each time you look at the value in the console, __repr__ gets called on some of the objects in the trouble array.  If you vary the size/shape of the array, you'll get different __repr__ calls (like just the first 6 for a long array, etc).

This behavior stops once the variable explorer is fully closed.  The __repr__ calls still go out if the variable explorer is not visible (unfocused tab) but still open.

And it really becomes a problem when the representations of the objects you are actually using start getting complicated.  And depending on how you configure your numpy print settings and how you organize your data, it is possible that quite a lot of expensive __repr__ calls go out every time you try to print something.

The given example has quite a simple __repr__, so doesn't replicate the slowdown.  But using my actual classes and data, I found the problem to go away simply by disabling __repr__ for my more complicated structures.  

But for now, the workaround is simply to close the variable explorer fully.
